win_hyperv_guest: New Windows Hyper-V module

wilmardo opened the pull request:
##### SUMMARY
Ansible Powershell for Hyper-V.
Supports most of the default operations and everything is idempotent.
Wrote test to support this claim ;) 

Design decisions:
* Network Adapter names and Internal Switch names should be unique otherwise the module does not know which one to assign.
* Tried to keep all operations in separate functions (New, Edit, Remove) for each individual component. The logic what to decide is in the `New-Guest` and `Edit-Guest` functions. Except for the power operations, those are decided in the main.
* `get_args` are passed to each get operation
* `set_args` are passed to each set operation

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
win_hyperv_guest

##### ADDITIONAL INFORMATION
Caveats:
 * Haven't been able to test this against a remote Hyper-V host (hostserver option). Also might be better to remove it completely? Somewhat obscure usecase to run Ansible on a host but letting it connect to a different host in the process?
* I think some of the pre-flight checks done in the module could be handled by the AnsibleModule options but I haven't wrapped my head arround it yet, suggestions welcome :)

Happy reviewing and a good 2019! 🎆

wilmardo commented:
Added tests for scaling CPU and memory. Live scaling is not possible, even on a generation 2 VM. This is due the system not being booted in a OS. But I think this current test is enough, the started VM requirement is from Hyper-V itself, these test prove that the module works correctly through the Cmdlet.

/ready_for_review again 🎉

wilmardo commented:
ready_for_review

Will try to join the IRC one of these days, need to fire up an new broker :)

wilmardo commented:
bot_status

it-praktyk commented:
Maybe nested virtualization can be used to perform integration tests?
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/nested-virtualization

wilmardo commented:
@it-praktyk I try to enable nested virtualization by installing Hyper V on the shippable host but, it complains about missing processor capabilities, see [ur](https://github.com/ansible/ansible/pull/50417#issuecomment-451197105)

wilmardo committed:
Adds win_hyperv_guest module

wilmardo commented:
Rebased against latest devel, also squashed into one commit. This issue about the tests is still standing, I have no clue how to get this running on Shippable.

jhawkesworth commented:
@wilmardo I suspect only way to get this running in Shippable would be to change the hosts used - having examined the error here: https://app.shippable.com/github/ansible/ansible/runs/122012/30/tests
(specifically 
```text
"Failed to install Windows Feature: Hyper-V cannot be installed The processor on this computer is not compatible with Hyper-V. To install this role, the processor must have a supported version of hardware-assisted virtualization, and that feature must be turned on in the BIOS."
I suggest asking in #ansible-devel irc channel to ask the ansible core developers if this is something that could be enabled.
Alternatively it may be necessary to disable the tests for this module from running in CI.  I don't know off the top of my head how this is done, but I suspect there wil be examples.  Again, worth asking in #ansible-devel irc channel, during USA business hours.
Hope this helps,
Jon

mattclay commented on test/integration/targets/win_hyperv_guest/aliases:
```diff
@@ -0,0 +1 @@
+shippable/windows/group4
```
Unfortunately we won't be able to run these tests on Shippable. Our Windows instances are run on EC2, which doesn't support nested virtualization. Instead, mark these tests as unsupported to prevent them from running in CI:

```suggestion
unsupported
```

wilmardo committed:
Update integration tests to be unsupported

Co-Authored-By: Matt Clay <matt@mystile.com>

wilmardo committed:
Update documentation with parameter types

zcuneo commented:
@wilmardo Thank you for all your work on this - it will certainly be a valuable module to have in Ansible!

So far, I've run into one issue related to Network Adapters. The error is only thrown if I add a network_adapters item with switch_name defined. If I simply give a name, the error is not thrown. Interestingly, if I update the network_adapters item with a switch_name later, it attaches to the virtual switch as expected.

`System.Management.Automation.ParameterBindingValidationException: Cannot validate argument on parameter 'VMNetworkAdapter'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again. ---> System.Management.Automation.ValidationMetadataException: The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
   at System.Management.Automation.ValidateNotNullOrEmptyAttribute.Validate(Object arguments, EngineIntrinsics engineIntrinsics)
   at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter, CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags)
   --- End of inner exception stack trace ---
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
fatal: [192.168.1.101]: FAILED! => {
    "changed": true,
    "invocation": {
        "module_args": {
            "cpu": 2,
            "disks": null,
            "force_stop": false,
            "generation": 2,
            "hostserver": null,
            "memory": "512MB",
            "name": "Test",
            "network_adapters": [
                {
                    "name": "test1",
                    "switch_name": "VM LAN"
                },
                {
                    "name": "test2",
                    "switch_name": "Bridge"
                }
            ],
            "secure_boot": false,
            "secure_boot_template": null,
            "state": "present",
            "timeout": 30,
            "wait_for_ip": false
        }
    },
    "msg": "Failed to update network adapter"
}
`

Thanks again for creating this handy module!

wilmardo commented:
@zcuneo Thanks for using the module! I will look into your issue as soon as I find time :)
Might take up to a month since my devserver is down at the moment, will try to reproduce your issue locally.

zcuneo commented:
@wilmardo Thank you for your response. After doing some homework on DSC, it seems the win_dsc module with xHyper-V might suit our needs just fine, but that made me wonder: what benefits would the dedicated win_hyperv_guest module bring vs DSC/xHyper-V?

jborean93 commented:
Hi!

Sorry for the extreme delay on getting to this PR but thank you very much for your submission to Ansible. It sincerely means a lot to us that you've taken the time to contribute.

While the functionality here in this new module would be useful for some people at this time we are not prepared to include it in the Ansible repository.

However, we recommend looking into providing this functionality through Ansible Galaxy via Ansible Collections. You can find more information about collections at https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html

Thank you once again for this and your interest in Ansible!

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
